src/imitation/algorithms/dagger.py

```diff
@@ -362,9 +346,16 @@ def save_trainer(self) -> str:
         saved_attrs = {
             attr_name: getattr(self, attr_name) for attr_name in self.SAVE_ATTRS
         }
+        # TODO(scottemmons): I removed what the tf version called `variable_values`.
+        #  Now, the two parts of the snapshot are nonoverlapping; both are needed fully
+        #  to reconstruct the trainer.
+        #  Do we want to stick with this design decision? Or was it important that the
+        #  previous version of the trainer object could be loaded solely from a single
+        #  snapshot file?
+        #  Furthermore, the tf version seemed to capture the Adam optimizer state.
+        #  Do we want this version to capture the optimizer state?
```

Comment:
No it's not necessary any more. We can move to a much simpler design now that we've got PyTorch. Possibly just doing `th.save(self, …)` is fine?

(also I don't think it's critically important to save optimiser state, but if we can do that at no cost then it would be nice)

Reply:
I've refactored so that it's not just doing `th.save(self, ...)` as you recommended. The refactoring involved changing some local functions to classes because the local functions couldn't be pickled.